docs/design/build-system.md

```diff
@@ -34,7 +67,8 @@ During compilation we also perform a number of code transformations using TypeSc
 
 AOT compilation requires loading HTML and CSS resources, referenced on Angular Components, as standalone strings with no external dependencies.
 However, Webpack compilations operate on the basis of modules and references between them.
-To obtain the standalone string we compile resources using a separate Webpack child compilation then extract the results.
+It's not possible to get the full compilation result for a given resource before the end of the compilation in webpack.
+To obtain the standalone string for a HTML/CSS resource we compile it using a separate Webpack child compilation then extract the results.
```

Comment:
Hopefully addresses https://github.com/angular/angular-cli/pull/16400#discussion_r356710447.

Reply:
@dgp1130 does this answer the question? Let me know if it's still not very clear. I think this is a rather complicated topic.

Reply:
So child compilations are necessary to get all the information at the end of a build. From that, my next question is "Why do we need a full compilation result for intermediate resources"? What info does the result provide which is needed in the middle of the pipeline?

Reply:
That's this bit:
> AOT compilation requires loading HTML and CSS resources, referenced on Angular Components, as standalone strings with no external dependencies.

Reply:
So it's impossible to load a file as a standalone string with no dependencies without using child compilations? That feels like something that should be possible in the regular compilation, unless I'm missing something?

Reply:
Not within a webpack plugin. Child compilations are kind of the standard method for these kind of operations.  Webpack also outputs JS and not a standalone string.  So the resource manager in the AOT plugin needs to execute the child compilation output in a separate VM context to get the actual string content for the resource ([code](https://github.com/angular/angular-cli/blob/0a83923f912c5514aee4789f0585f162f13495af/packages/ngtools/webpack/src/resource_loader.ts#L156)).

Reply:
It's not possible in webpack because webpack has a phased compilation. AOT compilation needs the final string (with all optimizations, and no module references) for CSS/HTML resources while compiling TS. But at that point webpack is still in the loading all resources, and hasn't gotten to the optimization phase.

If we were to get the resource string at that point during the loading phase it would be both unoptimized, and referencing other resources as well (if using imports).

Reply:
Ok, I think I see. So we need the fully optimized and bundled HTML/CSS resources, but those operations don't happen until the end of the compilation because they inherently require all the files to be processed. However, since we know they won't depend on other TS artifacts, we can run a child compilation which does not include TS files to let it compile the HTML/CSS files in their entirety, and use those during the TS compilations.

Does this sound like an accurate assessment?

Reply:
Yes that is accurate. Except the part about not including TS files. Child compilations are still full compilations and can do (most) anything.

For instance the web worker support uses a child compilation for building the web worker separately. Not really because we designed it that way, but because https://github.com/GoogleChromeLabs/worker-plugin uses child compilations and we support TS workers.

Reply:
Right, I understand it can include TS files. I was saying that it was choosing not to in that particular case. This seems reasonable to me.